packages/runtime/test-runtime-utils/src/mocksForRebasing.ts

```diff
@@ -0,0 +1,155 @@
+
+		// Messages in the same batch will have the same clientSequenceNumber
+		return this.clientSequenceNumber;
+	}
+
+	private submitInternal(message: ITrackableMessage) {
+		message.timesSubmitted++;
+
+		const metadata = { opId: message.opId, timesSubmitted: message.timesSubmitted };
+		this.factory.pushMessage({
+			clientId: this.clientId,
+			clientSequenceNumber: this.clientSequenceNumber,
+			contents: message.content,
+			referenceSequenceNumber: this.referenceSequenceNumber,
+			type: MessageType.Operation,
+			metadata,
+		});
+		this.addPendingMessage(
+			message.content,
+			{ ...(message.localOpMetadata as object), ...metadata },
+			this.clientSequenceNumber,
+		);
+	}
+
+	public rebase() {
+		this.currentBatch.forEach((message) => this.submitInternal(message));
+	}
+}
+
+/**
+ * To help debugging eventual consistency tests, all ops produced by this mock
+ * can be tracked using an unique id. The tracking information will also be included
+ * in the message metadata and local op metadata.
+ */
+interface ITrackableMessage {
+	/**
+	 * Message content
+	 */
+	content: any;
+	/**
+	 * local op metadata
+	 */
+	localOpMetadata: unknown;
+	/**
+	 * Unique identifier
+	 */
+	opId: string;
+	/**
+	 * How many times has this op been resubmitted
+	 */
+	timesSubmitted: number;
+}
+
+/**
+ * Specialized implementation of MockContainerRuntimeFactory for testing op rebasing.
+ */
+export class MockContainerRuntimeFactoryForRebasing extends MockContainerRuntimeFactory {
+	public createContainerRuntime(
+		dataStoreRuntime: MockFluidDataStoreRuntime,
+		overrides?: { minimumSequenceNumber?: number },
+	): MockContainerRuntimeForRebasing {
+		const containerRuntime = new MockContainerRuntimeForRebasing(
+			dataStoreRuntime,
+			this,
+			overrides,
+		);
+		this.runtimes.push(containerRuntime);
+		return containerRuntime;
+	}
+
+	private processMessage() {
+		if (this.messages.length === 0) {
+			throw new Error("Tried to process a message that did not exist");
+		}
+
+		// Explicitly JSON clone the value to match the behavior of going thru the wire.
+		const message = JSON.parse(
+			JSON.stringify(this.messages.shift()),
+		) as ISequencedDocumentMessage;
+
+		this.minSeq.set(message.clientId, message.referenceSequenceNumber);
+		// Messages from the same batch have the same sequence number
+		message.sequenceNumber = this.sequenceNumber;
+		message.minimumSequenceNumber = this.getMinSeq();
+		for (const runtime of this.runtimes) {
+			runtime.process(message);
+		}
+	}
+
+	public processOneMessage() {
+		// Increase the sequence number between batches
+		this.sequenceNumber++;
+		this.processMessage();
+	}
+
+	public processSomeMessages(count: number) {
+		// Increase the sequence number between batches
+		this.sequenceNumber++;
+		for (let i = 0; i < count; i++) {
+			this.processMessage();
+		}
+	}
+
+	public processAllMessages() {
+		// Increase the sequence number between batches
+		this.sequenceNumber++;
```

Comment:
Not clear why this.sequenceNumber is incremented differently if I call this API vs. call processOneMessage() multiple times